dotcom-rendering/src/components/AudioPlayer/AudioPlayer.tsx

```diff
@@ -263,6 +242,12 @@ export const AudioPlayer = ({
 
 		const audio = audioRef.current;
 
+		const onPause = () => setIsPlaying(false);
+		const onEnded = () => reportAudioEvent(mediaId, 'end');
+
+		const onWaiting = () => setIsWaiting(true);
+		const onCanPlay = () => setIsWaiting(false);
```

Comment:
I'm assuming (could be wrong) that the audio player will only be instantiated once so these functions are only created once removing the need to wrap in useCallback.

Reply:
> i think there's a couple of potential dangers here e.g. a page could have a playlist of URLs to play, in which case the mediaId would change, re-running the useEffect.

I think if we are using the same component (not multiple components) and changing the mediaId then either way those functions would (correctly) get re-created either via usecallbacks or inline here.

> is useCallback a bad pattern?

It isn't 'bad' per se but they have a creation cost in themselves which is potentially more expensive than just creating the functions each time. In this case because the player will re-render every second and we have lots of listener functions I'd expect they are saving some computation.

The case for removing them where possible is a) to do less work b) dependency arrays in hooks can be confusing and easy to over-trigger if mistakes are made in the referential equality of the items in the array.

https://kentcdodds.com/blog/usememo-and-usecallback

This is just a suggestion and an optimisation at that so we can always revisit if there is noticeable lag.